src/ac_power_flow_jacobian.jl

```diff
@@ -521,3 +521,139 @@ function calculate_loss_factors(
+- `time_step::Integer`: The calculated time step.
+"""
+function _calculate_voltage_stability_factors(
+    data::ACPowerFlowData,
+    J::ACPowerFlowJacobian,
+    time_step::Integer,
+)
+    ref, pv, pq = bus_type_idx(data, time_step)
+    pvpq = [pv; pq]
+    npvpq = length(pvpq)
+    rows, cols = block_J_indices(pvpq, pq)
+    σ, u, v = find_sigma_uv(J.Jv[rows, cols], npvpq)
+    data.voltage_stability_factors[ref, time_step] .= 0.0
+    data.voltage_stability_factors[first(ref), time_step] = σ
+    data.voltage_stability_factors[pv, time_step] .= 0.0
+    data.voltage_stability_factors[pq, time_step] .= v
+    return
+end
+
+"""
+    block_J_indices(data::ACPowerFlowData, time_step::Int) -> (Vector{Int32}, Vector{Int32})
+    
+Get the indices to reindex the Jacobian matrix from the interleaved form to the block form:
+
+| dP_dθ | dP_dV |
+| dQ_dθ | dQ_dV |
+
+# Arguments
+- `pvpq::Vector{Int32}`: Indices of the buses that are PV or PQ buses.
+- `pq::Vector{Int32}`: Indices of the buses that are PQ buses.
+
+# Returns
+- `rows::Vector{Int32}`: Row indices for the block Jacobian matrix.
+- `cols::Vector{Int32}`: Column indices for the block Jacobian matrix.
+"""
+function block_J_indices(pvpq::Vector{<:Integer}, pq::Vector{<:Integer})
+    rows = vcat(2 .* pvpq .- 1, 2 .* pq)
+    cols = vcat(2 .* pvpq, 2 .* pq .- 1)
+
+    return rows, cols
+end
+
+"""
+    find_sigma_uv(J::SparseMatrixCSC{Float64, Int32}, v_ix::Vector{Integer}, d_ix::Vector{Integer}; tol::Float64=1e-6, max_iter::Integer=100)
+
+Estimate the smallest singular value `σ` and corresponding left and right singular vectors `u` and `v` of a sparse matrix `G_s` (a sub-matrix of `J`).
+This function uses an iterative method involving LU factorization of the Jacobian matrix to estimate the smallest singular value of `G_s`. 
+The algorithm alternates between updating `u` and `v`, normalizing, and checking for convergence based on the change in the estimated singular value `σ`.
+The function uses the method described in the following publication:
+
+    P.-A. Lof, T. Smed, G. Andersson, and D. J. Hill, "Fast calculation of a voltage stability index," in IEEE Transactions on Power Systems, vol. 7, no. 1, pp. 54-64, Feb. 1992, doi: 10.1109/59.141687.
+
+# Arguments
+- `J::SparseMatrixCSC{Float64, Int32}`: The sparse Jacobian matrix.
+- `v_ix::Vector{Integer}`: Indices in the right singular vector `v` to be set to zero - corresponding to bus angles.
+- `d_ix::Vector{Integer}`: Indices in the left singular vector `u` to be set to zero - corresponfing to active power equations.
+
+# Keyword Arguments
+- `tol::Float64=1e-6`: Convergence tolerance for the iterative algorithm.
+- `max_iter::Integer=100`: Maximum number of iterations.
+
+# Returns
+- `σ::Float64`: The estimated smallest singular value.
+- `u::Vector{Float64}`: The estimated left singular vector.
+- `v::Vector{Float64}`: The estimated right singular vector.
+"""
+function find_sigma_uv(
```

Comment:
This function needs names for the variables... I can't tell what's happening here to make a meaningful review.

Reply:
add comment "in the paper it is ..."